File: zopyx/xform/ast.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Tuple


@dataclass
class Module:
    functions: dict
    rules: dict
    vars: dict
    namespaces: dict
    imports: list
    expr: "Expr | None"
    version: str = "2.0"


class Expr:
    pass


@dataclass
class Literal(Expr):
    value: object


@dataclass
class VarRef(Expr):
    name: str


@dataclass
class IfExpr(Expr):
    cond: Expr
    then_expr: Expr
    else_expr: Expr


@dataclass
class LetExpr(Expr):
    name: str
    value: Expr
    body: Expr


@dataclass
class ForExpr(Expr):
    name: str
    seq: Expr
    where: Optional[Expr]
    body: Expr


@dataclass
class MatchExpr(Expr):
    target: Expr
    cases: List[Tuple["Pattern", Optional[Expr], Expr]]
    default: Optional[Expr]

    def __post_init__(self):
        normalized = []
        for case in self.cases:
            if len(case) == 2:
                pattern, body = case  # type: ignore[misc]
                normalized.append((pattern, None, body))
            else:
                normalized.append(case)
        self.cases = normalized


@dataclass
class FuncCall(Expr):
    name: str
    args: List[Expr] = field(default_factory=list)
    named_args: List[Tuple[str, Expr]] = field(default_factory=list)


@dataclass
class ApplyExpr(Expr):
    expr: Expr
    ruleset: Optional[str] = None


@dataclass
class UnaryOp(Expr):
    op: str
    expr: Expr


@dataclass
class BinaryOp(Expr):
    op: str
    left: Expr
    right: Expr


@dataclass
class PathExpr(Expr):
    start: "PathStart"
    steps: List["PathStep"]


@dataclass
class Constructor(Expr):
    name: str
    attrs: List[Tuple[str, Expr]]
    contents: List["Content"]


@dataclass
class TextConstructor(Expr):
    expr: Expr


@dataclass
class CommentConstructor(Expr):
    expr: Expr


@dataclass
class PIConstructor(Expr):
    target: Expr
    value: Expr


@dataclass
class Text(Expr):
    value: str


@dataclass
class Interp(Expr):
    expr: Expr


Content = Expr


@dataclass
class PathStart:
    kind: str  # 'context', 'root', 'desc', 'desc_root', 'var', 'attr'
    name: Optional[str] = None


@dataclass
class PathStep:
    axis: str  # 'child', 'desc', 'self', 'parent', 'attr', 'desc_or_self'
    test: "StepTest"
    predicates: List[Expr]


@dataclass
class StepTest:
    kind: str  # 'name', 'wildcard', 'text', 'node', 'element', 'comment', 'pi', 'document'
    name: Optional[str] = None


class Pattern:
    pass


@dataclass
class WildcardPattern(Pattern):
    pass


@dataclass
class ElementPattern(Pattern):
    name: str
    var: Optional[str] = None
    child: Optional[Pattern] = None
    children: List[Pattern] = field(default_factory=list)
    attrs: List[Tuple[str, Optional[Literal]]] = field(default_factory=list)

    def __post_init__(self):
        if self.child is not None and not self.children:
            self.children = [self.child]
            self.child = None


@dataclass
class TypedPattern(Pattern):
    kind: str  # 'node', 'element', 'text', 'comment', 'pi', 'document'


@dataclass
class AttributePattern(Pattern):
    name: str
    value: Optional[Literal] = None


@dataclass
class LiteralPattern(Pattern):
    value: str


@dataclass
class Param:
    name: str
    type_ref: str | None = None
    default: Expr | None = None


@dataclass
class FunctionDef:
    params: List[Param]
    body: Expr


@dataclass
class RuleDef:
    pattern: Pattern
    body: Expr

File: xform-go/ast.go
package xform

type Module struct {
	Functions  map[string]FunctionDef
	Rules      map[string][]RuleDef
	Vars       map[string]Expr
	Namespaces map[string]string
	Imports    [][2]*string
	Expr       Expr
}

type Expr interface{}

type Literal struct{ Value any }

type VarRef struct{ Name string }

type IfExpr struct {
	Cond     Expr
	ThenExpr Expr
	ElseExpr Expr
}

type LetExpr struct {
	Name  string
	Value Expr
	Body  Expr
}

type ForExpr struct {
	Name  string
	Seq   Expr
	Where Expr
	Body  Expr
}

type MatchExpr struct {
	Target  Expr
	Cases   []MatchCase
	Default Expr
}

type MatchCase struct {
	Pattern Pattern
	Expr    Expr
}

type FuncCall struct {
	Name      string
	Args      []Expr
	NamedArgs []NamedArg
}

type NamedArg struct {
	Name string
	Expr Expr
}

type ApplyExpr struct {
	Expr    Expr
	Ruleset *string
}

type UnaryOp struct {
	Op   string
	Expr Expr
}

type BinaryOp struct {
	Op    string
	Left  Expr
	Right Expr
}

type PathExpr struct {
	Start PathStart
	Steps []PathStep
}

type Constructor struct {
	Name     string
	Attrs    []AttrConstructor
	Contents []Expr
}

type AttrConstructor struct {
	Name string
	Expr Expr
}

type TextConstructor struct{ Expr Expr }
type CommentConstructor struct{ Expr Expr }
type PIConstructor struct {
	Target Expr
	Value  Expr
}

type Text struct{ Value string }

type Interp struct{ Expr Expr }

type PathStart struct {
	Kind string
	Name *string
}

type PathStep struct {
	Axis       string
	Test       StepTest
	Predicates []Expr
}

type StepTest struct {
	Kind string
	Name *string
}

type Pattern interface{}

type WildcardPattern struct{}

type ElementPattern struct {
	Name     string
	Var      *string
	Child    Pattern
	Attrs    []PatternAttr
	Children []Pattern
}

type PatternAttr struct {
	Name  string
	Value *Literal
}

type TypedPattern struct{ Kind string }

type AttributePattern struct {
	Name  string
	Value *Literal
}

type LiteralPattern struct{ Value string }

type Param struct {
	Name    string
	TypeRef *string
	Default Expr
}

type FunctionDef struct {
	Params []Param
	Body   Expr
}

type RuleDef struct {
	Pattern Pattern
	Body    Expr
}
